README.md
# TCP-INT: Lightweight In-band Network Telemetry for TCP

In-band Network Telemetry (INT) provides visibility into the state of the network and can be used
for monitoring and debugging. However, existing INT implementations do not make telemetry available
to end-hosts. TCP-INT addresses this shortcoming by delivering network telemetry directly to the
end-host TCP stack, enabling end-hosts to correlate local TCP state with the state of the network
fabric.

TCP-INT is implemented in the TCP header as a new TCP option with three fields:

 - INTval: the link utilization (or queue depth if utilization is 100%).
 - HopID: the ID of most congested switch (the packet’s TTL at the switch).
 - HopLat: the sum of latencies experienced across all hops.

Each field has a corresponding echo-reply field (ecr) for the receiver to echo the telemetry back to
the sender.

### TCP-INT option format

```
 0                   1                   2                   3
 0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0 1
+---------------+---------------+-------+-------+---------------+
|  Kind = 0x72  |  Length = 12  |TagFreq|LinkSpd|    INTval     |
+---------------+---------------+-------+-------+---------------+
|  HopID=IP.TTL |                  HopLat (3B)                  |
+---------------+-------+-------+-------------------------------+
|     INTEcr    |LnkSEcr| HIDEcr|        HopLatEcr (2B)         |
----------------+-------+-------+-------------------------------+
```

### Workflow
- A process joins the `tcp-int` cgroup to indicate that its flows should be monitored by TCP-INT.
- When a packet is sent, the TCP-INT eBPF adds a TCP-INT header option to the TCP header.
- Upon receiving a packet with a TCP-INT header option, the switch updates the fields:
    - pkt.INTval = switch.INTval if switch.INTval > pkt.INTval
    - pkt.HopID = IP.TTL if switch.INTval > pkt.INTval
    - pkt.HopLat += latency through this switch
- The eBPF receives the telemetry and (possibly) sends it to user-space for consumption.
- When the ACK is sent, the eBPF sets the ecr fields to the latest INT received on the send path.


## Repo Organization

The TCP-INT repository is organized as follows:
```
code:
  ┣ src/bpf: eBPF module
  ┣ src/tools: tool for loading, managing and monitoring the module
  ┣ tcp-int-exporter: user-space agent to export INT to gRPC server
  ┗ scripts: helper scripts
```

### Switch Code

The switch-side P4 and control plane implementations are provided in a special release of Intel P4 Studio (aka Tofino SDE). To obtain access to the SDE, contact your Intel Sales Representative. Academic researchers can apply for access through the [Intel Connectivity Research Program (ICRP)](https://www.intel.com/content/www/us/en/products/network-io/programmable-ethernet-switch/connectivity-education-hub/research-program.html).


## Getting Started

### Prerequisites

* Linux Kernel 5.10.x or later

### Tested on

* Linux 5.11.0-17-generic and Linux 5.13.0-28-generic
* Ubuntu 21.04 and Ubuntu 21.10

### Setup steps for all end-hosts (clients and servers)

Examples in this README assume that you are running as the 'root' user. If not, then most commands will likely need privilege escalation using `sudo`.

#### 1. Install TCP-INT Dependencies

TCP-INT requires the following packages to be installed:
* build-essential
* clang
* llvm
* gcc-multilib
* libelf-dev
* linux-tools
* pkg-config
* libbpf 0.7 (See LibBPF setup below)
```
# For Debian based distributions
apt install -y build-essential clang llvm gcc-multilib libelf-dev linux-tools-$(uname -r) pkg-config
```

#### 2. Download and install Libbpf

More detailed installation instructions and dependencies can be found on the Libbpf Github page:<br>
https://github.com/libbpf/libbpf/tree/v0.7.0
```
wget https://github.com/libbpf/libbpf/archive/refs/tags/v0.7.0.tar.gz
tar xzf v0.7.0.tar.gz -C /opt
cd /opt/libbpf-0.7.0/src
make
LIBDIR=/lib/x86_64-linux-gnu make install
```
Note: User might need pkg-config to build libbpf.

#### 3. Copy TCP-INT source to end-hosts

Copy this source code repository to each end-host where you intend to use TCP-INT. For the purpose of these instructions we will assume it is extracted to `/opt/tcp-int`.

#### 4. Optional: update vmlinux.h file for the current kernel version

BPF programs need to hook into data structures in memory of the current running kernel, to do this they need an up-to-date header file containing information about the running kernel.

* Additional packages may be required to install the bpftool application

```
cd /opt/tcp-int/include/
bpftool btf dump file /sys/kernel/btf/vmlinux format c > vmlinux-$(uname -r).h
rm vmlinux.h
ln -s vmlinux-$(uname -r).h vmlinux.h
```

#### 5. Compile and deploy TCP-INT on the end-hosts

```
# Compile and install TCP-INT
cd /opt/tcp-int/src
make clean && make install
```
#### 6. Verify / set-up cgroups v2

TCP-INT requires cgroups v2 in order to work.
* The default 'load' command will attempt to create the following cgroup directory: `/sys/fs/cgroup/cgroup.tcp-int`
  * The default expects /sys/fs/cgroup to already exist as a cgroup2 mount.
* An alternative path can be specified when loading with the `-c` option.
  * For example: `tcp_int load -c /sys/fs/cgroup/unified/my-tcp-int.cgroup`

```
# Check if cgroups v2 are already mounted correctly
mount -l | grep cgroup2

# Example output 1: Ubuntu 21.04: Indicates that cgroups v2 is already mounted in a subdirectory of cgroup.
cgroup2 on /sys/fs/cgroup/unified type cgroup2 (rw,nosuid,nodev,noexec,relatime,nsdelegate)

# Example output 2: Debian 11 (and ubuntu 21.10): Indicates that cgroups v2 is mounted at the top level cgroup directory.
# In this case you can simply create the 'unified' subdirectory or define a custom tcp-int cgroup path using the -c option.
cgroup2 on /sys/fs/cgroup type cgroup2 (rw,nosuid,nodev,noexec,relatime,nsdelegate,memory_recursiveprot)
```

## Tagging policy
The TCP-INT eBPF inserts a TCP header option in outgoing packets. The switch can "tag" the packet by setting the INT fields in the header option. If the switch were to tag every packet, it could interfere with generic receive offload (GRO) in the Linux TCP stack, possibly hurting performance. To avoid this interference, the switch can tag *some* of the packets -- enough to provide fresh telemetry, without hurting performance.

### Tagging frequency
Instead of tagging every packet, the switch can tag packets at a given frequency. For example, if the tagging frequency is 16, one in every 16 packets will be tagged. The switch implements this logic probabilistically, so it's possible -- but unlikely -- for two successive packets to be tagged. The packets will be tagged in *expectation* at the specified tagging frequency.

To control the tagging frequency, the sender (eBPF) specifies the tagging frequency by setting the `tagfreqkey` field in the TCP header option. The switch uses `tagfreqkey` to decide whether to tag the packet.

### Default (static) tagging policy
By default, the tagging policy is to tag packets at a fixed tagging frequency. This tagging frequency is configured through the switch control plane. The sender requests the default tagging frequency by setting `tagfreqkey` to `TCP_INT_TAGFREQKEY_SWITCH_DEFAULT`.

### Dynamic tagging policy
Depending on network conditions and INT freshness requirements, the sender can dynamically change the tagging frequency. The current dynamic tagging policy identifies three states of a flow:

- `APPLIMITED`: the application isn't calling send() fast enough, so the in-flight packets are far below the CWND.
- `CONGESTED`: there's congestion in the network fabric.
- `UNCONGESTED`: there's no fabric congestion and the flow isn't application limited.

The thresholds for distinguishing these states, as well as the tagging frequency to use for each state, can be configured in `include/tcp_int_common.h`. The dynamic policy can be enabled through the `TCP_INT_ENABLE_DYNAMIC_TAGGING` define.

### Loading TCP-INT and running test applications

#### 1. Load TCP-INT on the end-hosts

This step should be run on all end-hosts each time that node is rebooted. This loads the TCP-INT BPF program.

```
# First, attempt to unload TCP-INT in case it was not left in a clean state
/usr/local/lib/bpf/tcp-int/tcp_int unload

# Now load the TCP-INT BPF program
/usr/local/lib/bpf/tcp-int/tcp_int load
```

#### 2. Disable TCP Timestamps (optional)

TCP timestamps may need to be disabled on BOTH the client and server end-hosts, depending on the current version of TCP-INT support on the switches. If this is necessary, run the following command on ALL end-host nodes:
```
# Disable TCP timestamps
sysctl -w net.ipv4.tcp_timestamps=0
```

#### 3. Server-side setup

The role of the server in TCP-INT test applications is to echo back the altered TCP-INT values to the client in acknowledgement (ACK) messages.

```
# Start bash session in tcp-int cgroup
bash
echo $$ >> /sys/fs/cgroup/cgroup.tcp-int/cgroup.procs

# Start server
pkill iperf3
iperf3 -s -p 5001
```

#### 4. Client-side setup

The role of the client is to initialise the TCP-INT TCP option and also to receive and process the echoed TCP-INT replies coming back from the server.

* Setting up a test stream with TCP-INT TCP Option initialised
```
# Open a shell terminal on the client node in the tcp-int cgroup
bash
echo $$ >> /sys/fs/cgroup/cgroup.tcp-int/cgroup.procs

# Generate a tcp stream using TCP-INT-eBPF
iperf3 -c <iPerf SERVER-IP> -p 5001 -w 20M -t 120
```

* Open a second console to control and observer TCP-INT
```
# Optionally disable TCP-INT echo on the client to reduce overheads
/usr/local/lib/bpf/tcp-int/tcp_int ecr-disable

# Start full tracing
/usr/local/lib/bpf/tcp-int/tcp_int trace
```
* Alternatively, display TCP-INT histograms showing in-network link utilisation and queue depth
```
# Optionally disable TCP-INT echo on the client to reduce overheads
/usr/local/lib/bpf/tcp-int/tcp_int ecr-disable

# Show TCP-INT histograms
watch -t -p -n 1 timeout -s SIGINT 1 "/usr/local/lib/bpf/tcp-int/tcp_int hist-int-perid | column"

# The tcp_int help text lists other modes of operation
/usr/local/lib/bpf/tcp-int/tcp_int help
```

## Running applications locally with network namespaces

This was tested an Ubuntu 21.10 VM with kernel `5.13.0-28-generic`.

#### 1. Load TCP-INT on the host
```
sudo /usr/local/lib/bpf/tcp-int/tcp_int load
```

#### 2. Setup the network namespaces
```
./code/scripts/namespaces.sh
```

#### 3. Start the application in the namespaces
* Open a shell for the server, add it to the tcp-int cgroup, and launch iperf3 in ns1:
```
bash
echo $$ | sudo tee -a /sys/fs/cgroup/cgroup.tcp-int/cgroup.procs
sudo ip netns exec ns1 iperf3 -s -p 5001
```

* Open another shell for the client and start the iperf3 client in ns2:
```
bash
echo $$ | sudo tee -a /sys/fs/cgroup/cgroup.tcp-int/cgroup.procs
sudo ip netns exec ns2 iperf3 -c 10.0.0.1 -p 5001 -t 120
```

#### 4. Add a `tc` pedit filter to mimic the switch
Modify TCP packets (except for SYN/FIN) sent from ns2 to port 5001: set the tcp-int option fields (intval=5, id=6).
This `tc` filter uses hard-coded offsets that **assume TCP timestamps are enabled**.
```
sudo ip netns exec ns2 tc qdisc replace dev v2 root handle 1: htb
sudo ip netns exec ns2 tc filter add dev v2 parent 1: protocol ip flower ip_proto tcp tcp_flags 0x0/0x7 dst_port 5001 action pedit munge offset 54 u8 set 5 continue
sudo ip netns exec ns2 tc filter add dev v2 parent 1: protocol ip flower ip_proto tcp tcp_flags 0x0/0x7 dst_port 5001 action pedit munge offset 56 u8 set 6 continue
```

#### 5. Monitor with tcp-int trace
```
sudo /usr/local/lib/bpf/tcp-int/tcp_int trace
```

#### 6. Remove the `tc` pedit filter
```
sudo ip netns exec ns2 tc filter del dev v2 parent 1:
```


changelog.md

# Changes to TCP-INT in version 0.4.0-alpha

## New features

 * Link speed
 * Available bandwidth

## Optimizations

 * New Version option to display currently running tcp-int version
 * Removed link utilzation feature

# Changes to TCP-INT in version 0.3.0-alpha

## New features

* Dynamic tagging policy: the eBPF specifies the tagging frequency to get timely updates without hurting performance
* New header option fields: TagFreqKey and LinkSpeed
* Compact encoding of Ecr fields: HopLat and IDEcr
* Performance evaluation scripts that configure the system and run workloads with various parameters
* Histograms of sent and received skb length
* Makefile configures the number of CPUs at compile time (instead of manually editing the header file)

## Optimizations

* Removed compiler warnings
* Avoid multiple eBPF map lookups by caching result

## Maintenance

* Renamed "SwitchID" to "HopID"
* Renamed "iratio" to "tagging frequency"

## Limitations

* The new 4-bit IDEcr encoding represents at most 15 hops on the path


# Changes to TCP-INT in version 0.2.0-alpha

## New features

* eBPF inserts TCP-INT option header into TCP packets being sent
* eBPF parses TCP-INT option header from received packets, and sends metrics to user-space for consumption
* Only programs (PIDs) in the tcp-int cgroup will have their TCP flows tagged by TCP-INT
* tcp_int user-space tool loads and configures eBPF
* User-space tcp_int tool prints trace of INT as it arrives
* The exporter receives INT events from the eBPF module and sends them as gRPC messages to a collector server
* Script and instructions for trying TCP-INT on a single server with network namespaces
* Script to plot timeseries of INT from trace generated by tcp_int tool

## Limitations

* Fine-grained flow filtering: currently, all TCP flows belonging to a PID in the tcp-int cgroup will be tagged. In addition, it would be useful to tag flows based on some filters (e.g., destination port, destination address, socket options).
* Testing: the tests in this release only exercise the basic functionality.


code/src/bpf/tcp_int.bpf.c
/* Copyright (C) 2021-2022 Intel Corporation */
/* SPDX-License-Identifier: GPL-2.0-only */

#include "vmlinux.h"
#include <bpf/bpf_helpers.h>
#include <bpf/bpf_endian.h>
#include "tcp_int_opt.h"
#include "tcp_int_common.h"
#include "tcp_int_common.bpf.h"
#include "bits.bpf.h"

/* TCP INT option definition */
#define TCP_INT_OPT_KIND 0x72
struct tcp_int_opt {
    __u8 kind;
    __u8 len;
    /* XXX bit-fields are sent in reverse on the wire: */
    unsigned linkspeed : 4;
    unsigned tagfreqkey : 4;
    tcp_int_val intval;
    tcp_int_id id;
    tcp_int_lat hoplat;
    tcp_int_val intvalecr;
    unsigned linkspeedecr : 4;
    unsigned idecr : 4;
    tcp_int_latecr hoplatecr;
} __attribute__((packed));

/* TCP INT configuration map */
#define TCP_INT_MAP_CONFIG_MAX_ENTRIES 16
struct {
    __uint(type, BPF_MAP_TYPE_HASH);
    __uint(max_entries, TCP_INT_MAP_CONFIG_MAX_ENTRIES);
    __type(key, __u16);
    __type(value, tcp_int_config_value);
} map_tcp_int_config SEC(".maps");

/* Perf events to user space */
struct {
    __uint(type, BPF_MAP_TYPE_PERF_EVENT_ARRAY);
    __uint(max_entries, TCP_INT_NUM_CPUS);
    __uint(key_size, sizeof(__u32));
    __uint(value_size, sizeof(__u32));
} map_tcp_int_events SEC(".maps");

/* Histograms */
struct {
    __uint(type, BPF_MAP_TYPE_HASH);
    __uint(max_entries, TCP_INT_HIST_TYPE_MAX);
    __type(key, enum tcp_int_hist_type);
    __type(value, struct tcp_int_hist);
} map_tcp_int_hists SEC(".maps");

/* Per hop-ID histograms */
struct {
    __uint(type, BPF_MAP_TYPE_HASH);
    __uint(max_entries, TCP_INT_MAX_PERID_HISTS);
    __type(key, enum tcp_int_hist_type);
    __type(value, struct tcp_int_hist_perid);
} map_tcp_int_hists_perid SEC(".maps");

static inline void tcp_int_enable_tcp_opt_cb(struct bpf_sock_ops *skops)
{
    int cb_flags;

    cb_flags = skops->bpf_sock_ops_cb_flags |
               BPF_SOCK_OPS_WRITE_HDR_OPT_CB_FLAG |
               BPF_SOCK_OPS_PARSE_UNKNOWN_HDR_OPT_CB_FLAG;
    bpf_sock_ops_cb_flags_set(skops, cb_flags);
}

static inline void tcp_int_reserve_hdr_opt(struct bpf_sock_ops *skops,
                                           __u32 size)
{
    bpf_reserve_hdr_opt(skops, size, 0);
}

static inline int tcp_int_is_mode_hist(void)
{
    int key = TCP_INT_CONFIG_KEY_HIST_ENABLE;
    tcp_int_config_value *ptr_enabled;

    ptr_enabled = bpf_map_lookup_elem(&map_tcp_int_config, &key);
    if (ptr_enabled)
        return ptr_enabled->u64;
    return 0;
}

static inline void tcp_int_update_hist(enum tcp_int_hist_type type,
                                       tcp_int_id hid, __u64 val, bool linear)
{
    struct tcp_int_hist_perid *hist_perid_ptr;
    struct tcp_int_hist *hist_ptr;
    __u64 slot;

    hist_perid_ptr = bpf_map_lookup_elem(&map_tcp_int_hists_perid, &type);
    hist_ptr = bpf_map_lookup_elem(&map_tcp_int_hists, &type);
    if (!hist_perid_ptr || !hist_ptr) {
        return;
    }

    if (hid >= TCP_INT_MAX_PERID_HISTS) {
        hid = TCP_INT_MAX_PERID_HISTS - 1;
    }

    slot = linear ? val : log2l(val);
    if (slot >= TCP_INT_HIST_MAX_SLOTS) {
        slot = TCP_INT_HIST_MAX_SLOTS - 1;
    }
    __sync_fetch_and_add(&(hist_perid_ptr->hist[hid].slots[slot]), 1);
    __sync_fetch_and_add(&hist_ptr->slots[slot], 1);
}

static inline void tcp_int_add_tcpopt(struct bpf_sock_ops *skops,
                                      struct tcp_int_state *istate)
{
    struct tcp_int_opt iopt = {0};
    __u64 val;

    iopt.kind = TCP_INT_OPT_KIND;
    iopt.len = sizeof(iopt);
    iopt.intvalecr = istate->intvalecr;
    iopt.idecr = istate->idecr;
    iopt.hoplatecr = istate->hoplatecr;
    iopt.linkspeedecr = istate->linkspeedecr;

#if TCP_INT_ENABLE_DYNAMIC_TAGGING
    if (skops->packets_out < 1)
        iopt.tagfreqkey = TCP_INT_TAGFREQKEY_APPLIMITED;
    else if (istate->qdepth > TCP_INT_CONG_QDEPTH_THRESH)
        iopt.tagfreqkey = TCP_INT_TAGFREQKEY_CONGESTED;
    else
        iopt.tagfreqkey = TCP_INT_TAGFREQKEY_UNCONGESTED;
#else
    iopt.tagfreqkey = TCP_INT_TAGFREQKEY_SWITCH_DEFAULT;
#endif // TCP_INT_ENABLE_DYNAMIC_TAGGING

    bpf_store_hdr_opt(skops, &iopt, sizeof(iopt), 0);

    if (tcp_int_is_mode_hist()) {
        val = skops->skb_len >> TCP_INT_SKBLEN_BITSHIFT;
        tcp_int_update_hist(TCP_INT_HIST_TYPE_TXSKBLEN, iopt.idecr, val, true);
    }
}

static void tcp_int_update_hists(struct bpf_sock_ops *skops,
                                 struct tcp_int_opt *iopt)
{
    enum tcp_int_hist_type type;
    bool linear;
    __u64 val;

    for (type = 0; type < TCP_INT_HIST_TYPE_MAX; type++) {
        switch (type) {
        case TCP_INT_HIST_TYPE_SRTT:
            val = skops->srtt_us >> 3;
            linear = false;
            break;
        case TCP_INT_HIST_TYPE_CWND:
            val = skops->snd_cwnd;
            linear = false;
            break;
        case TCP_INT_HIST_TYPE_AVAILBW:
            val = tcp_int_ival_to_avail_bw(iopt->intvalecr);
            linear = true;
            break;
        case TCP_INT_HIST_TYPE_QDEPTH:
            val = tcp_int_ival_to_qdepth(iopt->intvalecr) /
                  TCP_INT_BYTES_IN_KBYTE;
            linear = false;
            break;
        case TCP_INT_HIST_TYPE_HID:
            val = iopt->idecr;
            linear = true;
            break;
        case TCP_INT_HIST_TYPE_HLAT:
            val = tcp_int_hoplatecr_to_ns(iopt->hoplatecr);
            linear = false;
            break;
        case TCP_INT_HIST_TYPE_RXSKBLEN:
            val = skops->skb_len >> TCP_INT_SKBLEN_BITSHIFT;
            linear = true;
            break;
        default:
            continue;
        }
        tcp_int_update_hist(type, iopt->idecr, val, linear);
    }
}

static void tcp_int_send_event(struct bpf_sock_ops *skops,
                               struct tcp_int_opt *iopt)
{
    struct tcp_int_event event = {0};

    event.ts_us = bpf_ktime_get_ns() / 1000;
    event.dport = bpf_ntohl(skops->remote_port);
    event.sport = skops->local_port;
    event.family = skops->family;
    event.saddr_v4 = skops->local_ip4;
    event.daddr_v4 = skops->remote_ip4;
    /* TODO add ipv6 support */
    event.snd_cwnd = skops->snd_cwnd;
    event.srtt_us = skops->srtt_us;
    event.rate_delivered = skops->rate_delivered;
    event.rate_interval_us = skops->rate_interval_us;
    event.mss = skops->mss_cache;
    event.lost_out = skops->lost_out;
    event.intval = iopt->intvalecr;
    event.hid = iopt->idecr;
    event.hoplat = iopt->hoplatecr;
    event.return_hoplat = be24tohl(iopt->hoplat.u24);
    event.segs_out = skops->segs_out;
    event.bytes_acked = skops->bytes_acked;
    event.total_retrans = skops->total_retrans;
    event.link_speed = iopt->linkspeedecr;
    bpf_perf_event_output(skops, &map_tcp_int_events, BPF_F_CURRENT_CPU, &event,
                          sizeof(event));
}

static inline int tcp_int_is_ecr_enabled(void)
{
    int key = TCP_INT_CONFIG_KEY_ECR_DISABLE;
    tcp_int_config_value *ptr_enabled;

    ptr_enabled = bpf_map_lookup_elem(&map_tcp_int_config, &key);
    if (ptr_enabled)
        return !(ptr_enabled->u64);
    return 1;
}

static inline int tcp_int_is_mode_trace(void)
{
    int key = TCP_INT_CONFIG_KEY_TRACE_ENABLE;
    tcp_int_config_value *ptr_enabled;

    ptr_enabled = bpf_map_lookup_elem(&map_tcp_int_config, &key);
    if (ptr_enabled)
        return ptr_enabled->u64;
    return 0;
}

static inline int tcp_int_is_enabled(void)
{
    int key = TCP_INT_CONFIG_KEY_GLOBAL_ENABLE;
    tcp_int_config_value *ptr_enabled;

    ptr_enabled = bpf_map_lookup_elem(&map_tcp_int_config, &key);
    if (ptr_enabled)
        return ptr_enabled->u64;
    return 0;
}

static inline void tcp_int_process_tcpopt(struct bpf_sock_ops *skops,
                                          struct tcp_int_state *istate)
{
    struct tcp_int_opt iopt = {};
    bool _tcp_int_is_enabled; /* Optimization: cache tcp_int_is_enabled() */
    int rv;

    iopt.kind = TCP_INT_OPT_KIND;

    rv = bpf_load_hdr_opt(skops, &iopt, sizeof(iopt), 0);
    if (rv != sizeof(iopt)) {
        return;
    }

    /* Request echo only if there is an update */
    if (iopt.id) {
        istate->intvalecr = iopt.intval;
        istate->idecr = tcp_int_id_to_idecr(iopt.id);
        istate->hoplatecr =
            tcp_int_hoplat_to_hoplatecr(be24tohl(iopt.hoplat.u24));
        istate->linkspeedecr = iopt.linkspeed;
        istate->pending_ecr = true;
    }

    /* Ignore local events with no updates */
    if (iopt.idecr == 0) {
        return;
    }

    /* qdepth = qdepth_old * 7/8 + qdepth_new * 1/7 */
    istate->qdepth -= (istate->qdepth >> 3);
    istate->qdepth =
        istate->qdepth + (tcp_int_ival_to_qdepth(iopt.intvalecr) >> 3);

    _tcp_int_is_enabled = tcp_int_is_enabled();
    if (_tcp_int_is_enabled && tcp_int_is_mode_hist()) {
        tcp_int_update_hists(skops, &iopt);
    }
    if (_tcp_int_is_enabled && tcp_int_is_mode_trace()) {
        tcp_int_send_event(skops, &iopt);
    }
}

SEC("sockops")
int tcp_int(struct bpf_sock_ops *skops)
{
    struct tcp_int_state *istate;

    istate = tcp_int_get_state(skops->sk);
    if (!istate)
        return 1;

    switch (skops->op) {
    case BPF_SOCK_OPS_ACTIVE_ESTABLISHED_CB:
    case BPF_SOCK_OPS_PASSIVE_ESTABLISHED_CB:
        tcp_int_enable_tcp_opt_cb(skops);
        break;
    case BPF_SOCK_OPS_HDR_OPT_LEN_CB:
        if (tcp_int_is_enabled() ||
            (tcp_int_is_ecr_enabled() && istate->pending_ecr)) {
            tcp_int_reserve_hdr_opt(skops, sizeof(struct tcp_int_opt));
        }
        break;
    case BPF_SOCK_OPS_WRITE_HDR_OPT_CB:
        if (tcp_int_is_enabled() ||
            (tcp_int_is_ecr_enabled() && istate->pending_ecr)) {
            tcp_int_add_tcpopt(skops, istate);
            istate->pending_ecr = false;
        }
        break;
    case BPF_SOCK_OPS_PARSE_HDR_OPT_CB:
        tcp_int_process_tcpopt(skops, istate);
        break;
    default:
        break;
    }

    return 1;
}

char _license[] SEC("license") = "GPL";
